### PR TITLE
ROC-2161: added new route

### PR DESCRIPTION
**Pull request by yarpo**
### JIRA link ###
https://tools.hmcts.net/jira/browse/ROC-2161

### Change description ###
new feature

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/features/response/views/timeline.njk [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        {{ form.errorFor('rows[' + index + '][description]') }}
+      </span>
+    {% endif %}
+
+    <label class="form-label-bold mobile-show" for="timeline-date-{{ index }}">{{ index + 1 }}. Date</label>
+    <span class="form-hint mobile-show {% if error1 %}form-control-error{% endif %}">For example, 1 May 2017 </span>
+    <input class="form-control timeline-column-1" type="text" id="timeline-date-{{ index }}" name="timeline-date-{{ index }}" value="{{ form.valueFor(rows[' + index + '][date]) }}">
+
+    <label for="timeline-description-{{ index }}" class="form-label-bold mobile-show {% if error2 %}form-control-error{% endif %}">What happened</label>
+    <span class="form-hint mobile-show">For example, you might have signed a contract</span>
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: translating

File: src/main/features/response/views/timeline.njk [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      </span>
+    {% endif %}
+    {% if error2 %}
+      <span class="error-message" id="{{ 'rows[' + index + '][description][label]' }}">
+        {{ form.errorFor('rows[' + index + '][description]') }}
+      </span>
+    {% endif %}
+
+    <label class="form-label-bold mobile-show" for="timeline-date-{{ index }}">{{ index + 1 }}. Date</label>
+    <span class="form-hint mobile-show {% if error1 %}form-control-error{% endif %}">For example, 1 May 2017 </span>
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: translating

File: src/main/features/response/views/timeline.njk [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  {% set error2 = form.errorFor('rows[' + index + '][description]').length > 0 %}
+  {% set displayTotalError = totalError and loop.index0 == 0 %}
+
+  <div class="timeline-row {% if error %} form-group-error {% endif %}">
+    {% if displayTotalError %}
+      <span class="error-message" id="rows[label]">{{ form.errorFor('rows') }}</span>
+    {% endif %}
+    {% if error1 %}
+      <span class="error-message" id="{{ 'rows[' + index + '][date][label]' }}">
+        {{ form.errorFor('rows[' + index + '][date]') }}
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: translating

File: src/main/features/response/views/timeline.njk [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  {% set totalError = form.errorFor('rows').length > 0 %}
+
+  {% set error1 = form.errorFor('rows[' + index + '][date]').length > 0 %}
+  {% set error2 = form.errorFor('rows[' + index + '][description]').length > 0 %}
+  {% set displayTotalError = totalError and loop.index0 == 0 %}
+
+  <div class="timeline-row {% if error %} form-group-error {% endif %}">
+    {% if displayTotalError %}
+      <span class="error-message" id="rows[label]">{{ form.errorFor('rows') }}</span>
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: translating

File: src/main/features/response/views/timeline.njk [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    {% endif %}
+    {% if error1 %}
+      <span class="error-message" id="{{ 'rows[' + index + '][date][label]' }}">
+        {{ form.errorFor('rows[' + index + '][date]') }}
+      </span>
+    {% endif %}
+    {% if error2 %}
+      <span class="error-message" id="{{ 'rows[' + index + '][description][label]' }}">
+        {{ form.errorFor('rows[' + index + '][description]') }}
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: translating

File: src/main/features/response/views/timeline.njk [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      </span>
+    {% endif %}
+
+    <label class="form-label-bold mobile-show" for="timeline-date-{{ index }}">{{ index + 1 }}. Date</label>
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: translating

File: src/main/features/response/views/timeline.njk [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+          </div>
+        </div>
+        <div class="timeline-rows">
+          {% set n = form.model.rows.length %}
+          {% if n %}
+            {% for row in form.model.rows %}
+              {{ timelineRow(loop.index0, form) }}
+            {% endfor %}
+          {% else %}
+            {{ timelineRow(0, form) }}
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: this seems yuck, can't we just initialise it to 4?

File: src/main/features/response/views/timeline.njk [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        </div>
+      </div>
+
+      <div id="add-event-container form-group">
+        {{ linkButton('Add another event', 'action[addRow]', 'button button-secondary') }}
+      </div>
+      <br>
+
+      <div class="panel panel-border-wide">
+        <p>{{ t('Your response will be sent to the claimant Jan Clark. A judge will not consider your response unless your
```
Comment (timja-kainos) [inline src/main/features/response/views/timeline.njk]: Jan Clark looks hardcoded

Comment (yarpo) [reply src/main/features/response/views/timeline.njk]: fixed

File: src/main/features/response/views/timeline.njk (hunk 7 above)
Comment (yarpo) [reply src/main/features/response/views/timeline.njk]: refactored

File: src/main/features/response/routes/timeline.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction): Promise<void> => {
+      const form: Form<TimelineBreakdown> = req.body
+
+      if (form.hasErrors()) {
+        renderView(form, res)
+      } else {
+        res.locals.user.responseDraft.document.timeline = form.model
+
+        await DraftService.save(res.locals.user.responseDraft, res.locals.user.bearerToken)
+        renderView(form, res)
```
Comment (yarpo) [inline src/main/features/response/routes/timeline.ts]: next page is still not built this is why I redirect to the same page.

File: src/main/features/response/form/models/timelineBreakdown.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+  @ValidateNested({ each: true })
+  rows: TimelineRow[]
+
+  constructor (rows: TimelineRow[] = TimelineBreakdown.initialRows()) {
+    this.rows = rows
+  }
+
+  static initialRows (): TimelineRow[] {
+    let rows: TimelineRow[] = []
```
Comment (timja-kainos) [inline src/main/features/response/form/models/timelineBreakdown.ts]: this could be nicer with a reduce I think.

Comment (timja-kainos) [reply src/main/features/response/form/models/timelineBreakdown.ts]: rather than having a mutable collection here

File: src/main/features/response/form/models/timelineBreakdown.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  appendRow () {
+    this.rows.push(TimelineRow.empty())
+  }
+
+  private deserializeRows (rows: any): TimelineRow[] {
+    if (!rows) {
+      return TimelineBreakdown.initialRows()
+    }
+
+    let timelineRows: TimelineRow[] = []
```
Comment (timja-kainos) [inline src/main/features/response/form/models/timelineBreakdown.ts]: do a map here

File: src/main/features/response/form/models/timelineBreakdown.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
+import { ValidateNested } from 'class-validator'
+
+import { Serializable } from 'models/serializable'
+import { TimelineRow } from 'response/form/models/timelineRow'
+
+export const INIT_ROW_COUNT: number = 4
+
+export class TimelineBreakdown implements Serializable<TimelineBreakdown> {
```
Comment (rlewan) [inline src/main/features/response/form/models/timelineBreakdown.ts]: I think just `Timeline` would be enough. `TimelineBreakdown` sounds weird as well.

Comment (rlewan) [reply src/main/features/response/form/models/timelineBreakdown.ts]: I can also see many similarities with `ClaimAmountBreakdown` 🤔  Maybe we could extract some common parts as a subclass/utility class?

Comment (yarpo) [reply src/main/features/response/form/models/timelineBreakdown.ts]: because we have `AmountBreakdown` (that has very similar logic) I wanted to keep naming consistent.

Comment (timja-kainos) [reply src/main/features/response/form/models/timelineBreakdown.ts]: They are different though, I don't think the name makes sense

File: src/main/features/response/form/models/timelineBreakdown.ts (hunk 11 above)
Comment (yarpo) [reply src/main/features/response/form/models/timelineBreakdown.ts]: I'll apply this solution to `ClaimAmountBreakdown.deserializeRows` in following PR as well.

File: src/test/features/response/form/models/timelineBrakedown.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+      const actual: TimelineRow[] = TimelineBreakdown.initialRows()
+
+      expect(actual.length).to.equal(4)
+      expectAllRowsToBeEmpty(actual)
+    })
+  })
+
+  describe('fromObject', () => {
+
+    [false, null, undefined, 0, '', NaN].forEach(value => {
```
Comment (rlewan) [inline src/test/features/response/form/models/timelineBrakedown.ts]: As far as I remember we wanted to avoid coding against `null` in TS code, so I guess it can be removed from this list?

File: src/test/features/response/form/models/timelineRow.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+describe('TimelineRow', () => {
+
+  describe('empty', () => {
+
+    it('should return empty instances of TimelineRow', () => {
+
+      const actual: TimelineRow = TimelineRow.empty()
+
+      expect(actual instanceof TimelineRow).to.eq(true)
```
Comment (rlewan) [inline src/test/features/response/form/models/timelineRow.ts]: You can use [`instanceOf`](http://chaijs.com/api/bdd/#method_instanceof) Chai assertion here.

Comment (rlewan) [inline src/test/features/response/form/models/timelineRow.ts]: You can use [`instanceOf`](http://chaijs.com/api/bdd/#method_instanceof) Chai assertion here.

File: src/test/features/response/form/models/timelineBrakedown.ts [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+      actual.appendRow()
+
+      expect(actual.rows.length).to.be.eq(INIT_ROW_COUNT + 1)
+    })
+  })
+})
+
+function expectAllRowsToBeEmpty (rows: TimelineRow[]) {
+  rows.forEach(item => {
+    expect(item instanceof TimelineRow).to.eq(true)
```
Comment (rlewan) [inline src/test/features/response/form/models/timelineBrakedown.ts]: You can use [`instanceOf`](http://chaijs.com/api/bdd/#method_instanceof) Chai assertion here.

File: src/test/features/response/form/models/timelineBrakedown.ts (hunk 13 above)
Comment (yarpo) [reply src/test/features/response/form/models/timelineBrakedown.ts]: ok

File: src/main/features/response/form/models/timelineBreakdown.ts (hunk 12 above)
Comment (yarpo) [reply src/main/features/response/form/models/timelineBreakdown.ts]: ok

Comment (yarpo) [reply src/main/features/response/form/models/timelineBreakdown.ts]: ok

File: src/main/features/response/form/models/timelineBreakdown.ts (hunk 10 above)
Comment (yarpo) [reply src/main/features/response/form/models/timelineBreakdown.ts]: What would tell about:

```
static initialRows (): TimelineRow[] {
    return Array(INIT_ROW_COUNT).fill(TimelineRow.empty())
}
```

?

File: src/test/features/response/form/models/timelineRow.ts (hunk 14 above)
Comment (yarpo) [reply src/test/features/response/form/models/timelineRow.ts]: done

File: src/main/features/response/form/models/timeline.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  private deserializeRows (rows: any): TimelineRow[] {
+    if (!rows) {
+      return Timeline.initialRows()
+    }
+
+    let timelineRows: TimelineRow[] = rows.map(row => new TimelineRow().deserialize(row))
+
+    for (let i = 0; i < INIT_ROW_COUNT - rows.length; i++) {
```
Comment (timja-kainos) [inline src/main/features/response/form/models/timeline.ts]: this should call `initialRows`

File: src/main/features/response/form/models/timelineRow.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import { IsNotBlank } from 'forms/validation/validators/isBlank'
+
+export class ValidationErrors {
+  static readonly DATE_REQUIRED: string = 'Enter date'
+  static readonly DESCRIPTION_REQUIRED: string = 'Enter description'
+}
+
+export class TimelineRow {
+
+  @ValidateIf(o => o.description !== undefined)
```
Comment (timja-kainos) [inline src/main/features/response/form/models/timelineRow.ts]: this looks wrong, the only validation on this is maximum 20 characters and only if its not blank

File: src/main/public/stylesheets/components/_timeline.scss [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+  margin-top: 15px;
+}
+
+.mobile-show,
+.mobile-table-status {
+  display: none;
+}
+
+@media screen and (max-width: 480px) {
+  .mobile-hide {
```
Comment (timja-kainos) [inline src/main/public/stylesheets/components/_timeline.scss]: this doesn't look related to timeline, more of a utils sort of thing

File: src/main/features/response/routes/timeline.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    Paths.timelinePage.uri,
+    FormValidator.requestHandler(Timeline, Timeline.fromObject, undefined, ['addRow']),
+    actionHandler,
+    ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction): Promise<void> => {
+      const form: Form<Timeline> = req.body
+
+      if (form.hasErrors()) {
+        renderView(form, res)
+      } else {
+        res.locals.user.responseDraft.document.timeline = form.model
```
Comment (timja-kainos) [inline src/main/features/response/routes/timeline.ts]: empty timeline rows are to be filtered out before saving, see A/C 4.

File: src/main/features/response/form/models/timeline.ts [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  readonly type: string = 'breakdown'
+
+  @ValidateNested({ each: true })
+  rows: TimelineRow[]
+
+  constructor (rows: TimelineRow[] = Timeline.initialRows()) {
+    this.rows = rows
+  }
+
+  static initialRows (): TimelineRow[] {
```
Comment (timja-kainos) [inline src/main/features/response/form/models/timeline.ts]: does this need to be public?

File: src/main/features/response/form/models/timelineRow.ts (hunk 17 above)
Comment (yarpo) [reply src/main/features/response/form/models/timelineRow.ts]: done

File: src/main/public/stylesheets/components/_timeline.scss (hunk 18 above)
Comment (yarpo) [reply src/main/public/stylesheets/components/_timeline.scss]: moved to _mobile.scss

File: src/main/features/response/form/models/timeline.ts [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    return this
+  }
+
+  appendRow () {
+    if (this.rows.length < MAX_NUMBER_OF_EVENTS) {
+      this.rows.push(TimelineRow.empty())
+    }
+  }
+
+  clearUselessRows () {
```
Comment (rlewan) [inline src/main/features/response/form/models/timeline.ts]: I think "Useless" is not the most accurate word here (https://translate.google.com/#en/pl/useless), what do you think about using "Blank" instead?

Comment (rlewan) [reply src/main/features/response/form/models/timeline.ts]: How about "[Superfluous](https://translate.google.com/#en/pl/superfluous)"?
